### ISSUE TITLE
Supabase CORS Error on REST API Requests: No 'Access-Control-Allow-Origin' header is present

### ISSUE THREAD

**Issue by caleb531** [supabase/supabase#7627]
# Bug report

## Describe the bug

I am using the Supabase SaaS service and supabase-js for my latest web application that I intend to launch soon. I am experiencing a CORS issue for new user accounts of my app, whereby a new user account receives a 400 CORS response code when calling my public DB tables via the REST API.

<img width="1158" alt="1-console" src="https://user-images.githubusercontent.com/872474/178053101-ae11e924-1eda-4853-8321-3ba19eba4535.png">

Upon further inspection, it appears that the `access-control-allow-origin` header is present on the OPTIONS response, but _not_ the GET/POST response. Conversely, for users where requests have completed successfully, the `access-control-*` headers are present on both preflight and non-preflight requests. This issue currently applies to both GET and POST requests.

<img width="1063" alt="2-options-200" src="https://user-images.githubusercontent.com/872474/178053778-93f5bda6-deed-470f-907b-328878ab3c70.png">

<img width="1042" alt="3-get-400" src="https://user-images.githubusercontent.com/872474/178053907-15e54e67-b328-4f4a-adb4-832cc4408e67.png">

The interesting thing is, once the CORS issue occurs for a new user account I just created, the CORS errors will persist for the lifetime of that user account. I've tried clearing cookies, localStorage, and even switched browsers—nothing resolves the CORS issue for an affected user.

If I clear browsing data and log into a different user I created previously, the 400 CORS errors go away while I am signed into that user account.

It's an issue I have been unable to reproduce locally, and I haven't been able to draw any inferences from my `auth.users` table on https://app.supabase.com/.

## Expected behavior

The GET/POST request should include CORS headers, and said requests should complete successfully for all users. I suspect this is an issue with Supabase platform's server-side, as I cannot control their response headers otherwise.

## Screenshots

See above.

## System information

- OS: macOS Monterey 12.4
- Browser: Tested on Google Chrome 103.0.5060.114 and Brave browser (v1.40.113 Chromium: 103.0.5060.114)
- Version of supabase-js: v1.35.4
- Version of Node.js: v16 (hosted on Vercel)

## Additional context

I've checked my API Logs in my Supabase control panel, but they don't reveal much.

<img width="975" alt="4-supabase-api-logs" src="https://user-images.githubusercontent.com/872474/178060388-39015c8e-58ab-470d-a1f2-fdd2351ed8a1.png">

**Comment by steve-chavez** [comment]
Could this be related to https://github.com/supabase/postgrest-js/issues/286#issuecomment-1165213335?

**Comment by caleb531** [comment]
@steve-chavez I saw that issue, and tried disabling all my browser extensions in both Brave and Chrome. No luck.

I also tried in Safari, which has no extensions enabled, and I receive the same CORS errors for that user account.

**Comment by GaryAustin1** [comment]
Is there anything in the Database Postgres logs?
The cors error maybe caused by whatever is causing the 400 error versus 400 caused by cors.  I don’t have a specific case with SB to point to, but have seen cors errors caused by server errors.  

Can you turn off RLS if you are using and make sure it is not failing in some bad way… although I think that would cause a different error like 500…

**Comment by caleb531** [comment]
@GaryAustin1 Addressing your questions in LIFO order 😉:

1. I disabled RLS on both of my endpoints, but I'm still getting the CORS errors, even after refreshing the page and signing out / back in (I have since re-enabled them after submitting this comment for security):

<img width="959" alt="Screen Shot 2022-07-08 at 5 10 58 PM" src="https://user-images.githubusercontent.com/872474/178083858-3cb171db-f01b-43c8-b8d4-77c199fb65c2.png">

2. Here are my Postgres logs. Nothing particularly stands out to me—they look mostly admin-related (e.g. `identity="supabase_admin"`

<img width="478" alt="Screen Shot 2022-07-08 at 5 08 42 PM" src="https://user-images.githubusercontent.com/872474/178083904-e6d15924-612b-46dd-abdf-33682fb84d38.png">

**Comment by burggraf** [comment]
Can you please try disabling the CAPTCHA to see if that resolves the issue?

**Comment by caleb531** [comment]
@burggraf Thank you for this. I think I have identified the root cause, and ask for your feedback on how to proceed from here:

## TL;DR

The CAPTCHA responses is inflating the size of the JWT, since the CAPTCHA responses get stored in `raw_user_meta_data` in `auth.users` table (or `user_metadata` on the frontend `User` object). And since the JWT is sent in every API request, Supabase fails because it can't seem to handle a request of that size.

## Details

Before stripping out the CAPTCHA functionality, 100% of new user accounts that I've created experienced the CORS error.

However, _after_ stripping out the CAPTCHA functionality (per your recommendation), new user accounts no longer experience the CAPTCHA error, yet user accounts that were created when the CAPTCHA was still deployed are still exhibiting the CORS error.

Going further, I examined the `raw_user_meta_data` on the newly-registered user, and noticed that the CAPTCHA responses are stored in this JSON object.

```sql
select raw_user_meta_data FROM auth.users WHERE id = <userid>
-- JSON object with first_name, last_name, h-captcha-response, and g-recaptcha-response keys returned (the latter two are quite large)
```

When I run the following `update` query and sign out / back in, the CORS errors disappear for that user:

```sql
update auth.users set raw_user_meta_data = '{"last_name":"Doe","first_name":"John"}' WHERE id = '<userid>'
-- the user now has captcha data removed from their raw_user_meta_data field
```

Given this, I have a hypothesis for what's happening here:

1. When the `raw_user_meta_data` object contained the CAPTCHA fields, it was about 8 KB in size
2. Since `raw_user_meta_data` gets encoded as part of the JSON web token, the length of the JWT gets inflated substantially
    1. This is consistent with what I've been seeing, as every user account affected by the CORS error has an `Authorization: Bearer` header of 12kb on its own (pretty substantial for a single header value)
3. Because the JWT is so large, the Superbase SaaS platform either refuses to handle it, or is simply unable to process a request that large

## Questions

1. Is it expected behavior for the hcaptcha responses to be stored in the `raw_user_meta_data` object on account creation? I would imagine these values can be evaluated once and then thrown away upon successful verification
2. If you agree with my findings, do you have any recommendations about how to proceed from here?

**Comment by caleb531** [comment]
Update: after reviewing my code, I figured out the issue—a mistake on my end. My Sign Up submission code was something like the following:

```ts
const fields = serializeForm(event.currentTarget);
return supabase.auth.signUp(
    {
        email: fields.email,
        password: fields.password
    },
    {
        captchaToken,
        data: _.omit(fields, [
            'email',
            'confirm_email',
            'password',
            'confirm_password'
        ])
    }
);
```

As you can see, I'm serializing all form values into a JSON object, and populating those into the user metadata store on account creation. However, while my `_.omit` call excludes irrelevant fields like email/password fields, it should _also_ be excluding the CAPTCHA fields. To solve, if I invert the logic to use `_.pick` instead of `_.omit` to be more explicit, then I should be good to go.

Therefore, I am closing this issue, as it does not represent an issue with Supabase itself. Sorry for any inconvenience, but thank you for everyone's time!

Caleb

**Comment by jonathanlal** [comment]
For me happened because my windows clock was not in sync - why I have no idea but ever since day light savings time my PC is starting 1h behind and gives me this error whenever the clock is wrong

**Comment by lorenaserbeco-del** [comment]
https://claude.ai
